src/lang/proc_macros/client/mod.rs

```diff
@@ -142,6 +150,7 @@ impl ProcMacroClient {
         match self.send_request_untracked::<M>(id, &params) {
             Ok(()) => {
                 requests_params.insert(id, map(params));
+                self.analysis_progress_tracker.register_procmacro_request();
```

Comment:
I believe that there is a small chance that a yet alive snapshot from previous diagnostics calculation was already in `db.get_attribute_expansion` before its cancellation. It will still call this method and possibly register proc macro request in analysis tracker AFTER `track_analysis` is called to start tracking the new diagnostics batch. It can cause us to never finish waiting for diagnostics in tests, since:
- the next batch won't be scheduled
- we will always think there is some proc macro yet to resolve while it may not be